src/loki/search.cc

```diff
@@ -385,102 +376,32 @@ struct bin_handler_t {
-          reaches.back() += reaches[reach_index] - 1;
-          reaches[reach_index] = reaches.back();
-          return;
-        }
-
-        // Increment and recurse
-        ++reaches.back();
-        size_t previous = reach_index;
-        depth_first(tile, n, reach_index);
-
-        // if we saw the edge in a previous run we want to be done completely
-        if (reach_index != previous) {
-          return;
-        }
-      }
-    }
-
-    // Follow transition to other hierarchy levels
-    if (node->transition_count() > 0) {
-      const NodeTransition* trans = start_tile->transition(node->transition_index());
-      for (uint32_t i = 0; reaches.back() < max_reach_limit && i < node->transition_count();
-           ++i, ++trans) {
-        const auto* trans_tile = reader.GetGraphTile(trans->endnode());
-        if (trans_tile == nullptr) {
-          // Protect against missing tile
-          continue;
-        }
-        const NodeInfo* n = trans_tile->node(trans->endnode());
-        if (!node_filter(n)) {
-          // try to mark the node
-          auto inserted = reach_indices.emplace(trans->endnode(), reach_index);
-          // we saw this one already
-          if (!inserted.second) {
-            // we've seen this node in this run so just skip it
-            if (reach_index == inserted.first->second) {
-              continue;
-            }
-
-            // signal the recursion to stop
-            reach_index = inserted.first->second;
-            // merge this paths reach with the previously found one
-            reaches.back() += reaches[reach_index] - 1;
-            reaches[reach_index] = reaches.back();
-            return;
-          }
-
-          // For transitions, recurse but don't increment so we don't double count nodes
-          size_t previous = reach_index;
-          depth_first(trans_tile, n, reach_index);
-
-          // if we saw the edge in a previous run we want to be done completely
-          if (reach_index != previous) {
-            return;
-          }
-        }
-      }
-    }
+  directed_reach get_reach(const DirectedEdge* edge) {
+    // if its in cache return it
+    auto itr = directed_reaches.find(edge);
+    if (itr != directed_reaches.cend())
+      return itr->second;
+
+    // notice we do both directions here because in the end we use this reach for all input locations
+    auto reach =
+        SimpleReach(edge, max_reach_limit, reader, kInbound | kOutbound, edge_filter, node_filter);
+    directed_reaches[edge] = reach;
+    return reach;
   }
 
   // do a mini network expansion or maybe not
-  unsigned int check_reachability(std::vector<projector_wrapper>::iterator begin,
-                                  std::vector<projector_wrapper>::iterator end,
-                                  const GraphTile* tile,
-                                  const DirectedEdge* edge) {
+  directed_reach check_reachability(std::vector<projector_wrapper>::iterator begin,
```

Comment:
here we actually decide if we should or shouldnt check the reach. there is an optimization in there where we say hey if we aren't going to actually get a better candidate by checking this candidates reach then dont waste our time, it seems in practice this really helps avoid any reach calculation at all. additionally even if we do get a better edge candidate we dont actually compute its reach if we already have reachable candidates.